### PR TITLE
fix: IOU - Disabled tag is greyed in list but disabled category is shown bold in list.

### PR DESCRIPTION
**Pull request by Krishna2323**
### Fixed Issues
$ https://github.com/Expensify/App/issues/36485
PROPOSAL: https://github.com/Expensify/App/issues/36485#issuecomment-1943884263

### Tests
1. Launch app as employee
2. Create a manual request > Enter amount > Select employee workspace chat
3. Enter merchant > Select any category > Create request
4. In admin account, disable the category used to create above expense
5. In employee account, in Workspace chat, open the newly created request to open details page.
6. Open category edit page by clicking on the category
7. Verify disabled category is visible
8. Search for the category and verify disabled category is appears in the search result with the checkmark
9. Change category to any other option > Search for the changed category 
10. Verify category is appears in the search result with the checkmark

- [x] Verify that no errors appear in the JS console

### Offline tests
1. Launch app as employee
2. Create a manual request > Enter amount > Select employee workspace chat
3. Enter merchant > Select any category > Create request
4. In admin account, disable the category used to create above expense
5. In employee account, in Workspace chat, open the newly created request to open details page.
6. Open category edit page by clicking on the category
7. Verify disabled category is visible
8. Search for the category and verify disabled category is appears in the search result with the checkmark
9. Change category to any other option > Search for the changed category 
10. Verify category is appears in the search result with the checkmark

### QA Steps
1. Launch app as employee
2. Create a manual request > Enter amount > Select employee workspace chat
3. Enter merchant > Select any category > Create request
4. In admin account, disable the category used to create above expense
5. In employee account, in Workspace chat, open the newly created request to open details page.
6. Open category edit page by clicking on the category
7. Verify disabled category is visible
8. Search for the category and verify disabled category is appears in the search result with the checkmark
9. Change category to any other option > Search for the changed category 
10. Verify category is appears in the search result with the checkmark

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/85894871/ddcbe525-ea93-47cf-bc3f-55d8898cdbc7

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/85894871/0792b4a6-0434-41f0-be40-8fb9abc30730

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/85894871/28745c33-98bf-4955-bfef-7fe1d3bd3a79

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/85894871/0c54c7a1-0123-41bf-a05e-bf063e138df9

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/85894871/88cecc41-6b55-4182-b067-ac4c6951ee75

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/85894871/6dac9460-7907-445d-8a5b-cd29fb73a281

</details>

### CODE REVIEW

**Comment by alitoshmatov** [comment]
Not sure what is the best approach here. @dangrous Can you take a look at [this](https://github.com/Expensify/App/pull/37397#issuecomment-1975867520). 

I am leaning towards modifying `SelectionList` to show items greyed when they are disabled. But it will affect whole app and change its design slightly

**Comment by dangrous** [comment]
Yeah I think we should probably grey it out, it isn't super clear if it's only the cursor style. We should just do a set of testing to see if there's anywhere else it shows up greyed out unexpectedly - I think overall it's a clearer UI, though.

**Comment by alitoshmatov** [comment]
@Krishna2323 Can you resolve conflicts

**Comment by alitoshmatov** [comment]
Thank you, I will update our next options acordingly

**Comment by dangrous** [comment]
Hrm I might bring in @Expensify/design to confirm what the expected UX for disabled categories/tags is. @alitoshmatov / @Krishna2323 can you clarify what currently happens (without any changes made here) to make sure it's correct?

And then yeah I think we can roll in that new issue you found.

**Comment by alitoshmatov** [comment]
Currently, you have money request and specific category and tag is selected for this. If you disabled that selected category and/or tag from OD and in ND open categories or tag in money request disabled tags appear as normal(previously they were grayed out and mouse was disabled). 

**Note:** Not disabling cursor kinda makes sense, since you can just click on that category or tag to unselect it without selecting new item

**Comment by dannymcclain** [comment]
Is this in any way related to [this convo](https://expensify.slack.com/archives/C03U7DCU4/p1709661920842379) we recently had in Slack about disabled categories?

Basically disabled categories (or tags) that are currently selected should look disabled, still have the checkmark, and the user should be able to interact with it _only to remove_ the selection. Once the selection has been removed, I don't think the disabled category (or tag) would show up in the list anymore. Does that sound right @shawnborton?

**Comment by alitoshmatov** [comment]
### For comparison    

Then: 
Screenshot 2024-03-13 at 8 55 54 PM

Now:
Screenshot 2024-03-13 at 8 58 17 PM

**Comment by shawnborton** [comment]
> Basically disabled categories (or tags) that are currently selected should look disabled, still have the checkmark, and the user should be able to interact with it only to remove the selection. Once the selection has been removed, I don't think the disabled category (or tag) would show up in the list anymore. Does that sound right @shawnborton?

That makes sense to me, though I guess we'd still show the selected BG color on the row too right?

**Comment by dannymcclain** [comment]
> though I guess we'd still show the selected BG color on the row too right?

Yeah I think we would.

**Comment by dangrous** [comment]
Are you clear on how to move forward @alitoshmatov ? Let us know and if not we can figure out how to clarify!

**Comment by dangrous** [comment]
Hm, yeah, I think we should probably update both. Do you mind posting in the [issue](https://github.com/Expensify/App/issues/36485) and we can involve the BZ team in any update on pricing?

**Comment by mallenexpensify** [comment]
[Commented on the issue](https://github.com/Expensify/App/issues/36485#issuecomment-2026286177), @Krishna2323 can you please keep this PR moving along, [another issue](https://github.com/Expensify/App/issues/38226) is held on it.  Thx

File: src/components/SelectionList/RadioListItem.tsx [hunk 1]
```diff
@@ -55,7 +55,8 @@ function RadioListItem({
                             styles.sidebarLinkTextBold,
                             isMultilineSupported ? styles.preWrap : styles.pre,
                             item.alternateText ? styles.mb1 : null,
-                            isDisabled && styles.colorMuted,
+                            /* eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing */
+                            (isDisabled || item.applyDisabledStyle) && styles.colorMuted,
```
Comment (alitoshmatov) [inline src/components/SelectionList/RadioListItem.tsx:59]: A little bit confusing logic, can we just change it so that we align with eslint rule
```suggestion
(isDisabled || item.applyDisabledStyle) ? styles.colorMuted : null,
```

File: src/components/SelectionList/types.ts [hunk 2]
```diff
@@ -73,6 +73,9 @@ type ListItem = {
     /** Whether this option is disabled for selection */
     isDisabled?: boolean | null;
 
+    /** To apply diabled style when item is not diabled to unselect */
+    applyDisabledStyle?: boolean | null;
```
Comment (alitoshmatov) [inline src/components/SelectionList/types.ts:77]: Why exactly we need separate `applyDisabledStyle` attribute. Can't we achieve the same result with `isDisabled`?

File: src/components/CategoryPicker.tsx [hunk 3]
```diff
@@ -71,7 +71,6 @@ function CategoryPicker({selectedCategory, policyCategories, policyRecentlyUsedC
     }, [policyRecentlyUsedCategories, debouncedSearchValue, selectedOptions, policyCategories]);
 
     const selectedOptionKey = useMemo(() => (sections?.[0]?.data ?? []).filter((category) => category.searchText === selectedCategory)[0]?.keyForList, [sections, selectedCategory]);
-
```
Comment (alitoshmatov) [inline src/components/CategoryPicker.tsx:74]: Unnecessary change

File: src/components/SelectionList/types.ts (hunk 2 above)
Comment (alitoshmatov) [reply src/components/SelectionList/types.ts:77]: After a lot of thinking and reviewing, I think this is not a good approach to this problem. However I didn't manage to come up with better solution yet. But I do think that we should really think about this solution before proceeding with it.

@dangrous What do you think about changes with `applyDisabledStyle`?

Comment (dangrous) [reply src/components/SelectionList/types.ts:77]: Hi! I'm still partially on leave. Let me give this some thought, see if I can figure out something!

Comment (dangrous) [reply src/components/SelectionList/types.ts:77]: Okay, well, I came up with nothing. Can you post about this in the Slack channel and see if anyone else has some clever ideas to cover all the situations? I also would love not to go this route since it seems messy - but if it's the only option, then we can do it.

Comment (dangrous) [reply src/components/SelectionList/types.ts:77]: I think that sounds better to me, what do you think @alitoshmatov? Alternately, can we use an existing value of `isSelected` or similar? How do we know if it's currently selected right now?

Comment (alitoshmatov) [reply src/components/SelectionList/types.ts:77]: I agree, can we just check if currently selected item is disabled, if yes we would allow it to be deselected without introducing new attribute @Krishna2323

Comment (dangrous) [reply src/components/SelectionList/types.ts:77]: Bump @Krishna2323 what do you think about this? Do you think you can work on the solution soon so we can get this updated? Thanks!

**Comment by alitoshmatov** [comment]
@Krishna2323 Can you check recording below, disabled tag is not showing even if it was selected
https://github.com/Expensify/App/assets/59907218/48534bef-d6b1-460f-b57f-1de324a607e0

File: src/components/SelectionList/RadioListItem.tsx [hunk 4]
```diff
@@ -52,6 +52,7 @@ function RadioListItem<TItem extends ListItem>({
                             styles.sidebarLinkTextBold,
                             isMultilineSupported ? styles.preWrap : styles.pre,
                             item.alternateText ? styles.mb1 : null,
+                            /* eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing */
```
Comment (alitoshmatov) [inline src/components/SelectionList/RadioListItem.tsx:55]: why do we need this

**Comment by alitoshmatov** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/59907218/a082e30a-7008-4427-b260-f242921d6439

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/59907218/799e5786-c811-4c73-864c-7033f667ec19


</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/59907218/ee5cf09a-a074-41c7-9971-4b12f961fb9a

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/59907218/4c497254-5502-4d1a-8c16-5c1d089fb721

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/59907218/af5176a1-7562-437e-959d-2fa46d5de8ae


</details>

**Review by alitoshmatov — APPROVED** [review]
Finally! I think we can merge it

**Review by dangrous — APPROVED** [review]
@Krishna2323 this looks good, but one question - can you explain the changes that you made to the tests? I get adding `isSelected: undefined` but not why the changes on lines 1070,1071,1239,1240.

Thank you!

**Comment by dangrous** [comment]
Cool! That makes sense to me.
